### DISCUSSION TITLE
Supported Python versions

### DISCUSSION THREAD

**Discussion by adamjstewart** [torchgeo/torchgeo#656] [General]
This discussion is about the versions of Python that TorchGeo should support. There are two commonly used support timelines:

* [CPython support cycle](https://endoflife.date/python): 5 years after minor release
* [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html): 3.5 years after minor release

To illustrate the difference, the latest version of CPython that is still supported is 3.7, and will continue to be supported for another 11 months. However, projects that have adopted NEP 29 have already dropped Python 3.7 and will drop 3.8 in another 9 months.

NEP 29 has achieved a high rate of adoption in the scientific Python ecosystem, with many projects dropping support for Python versions long before CPython does. I just noticed that PyTorch also recently adopted NEP 29 (https://github.com/pytorch/pytorch/issues/74203, https://github.com/pytorch/pytorch/pull/77304, https://github.com/pytorch/pytorch/issues/80513). In the past, we've been following the CPython support cycle because PyTorch did as well, but if PyTorch plans on following NEP 29 we may want to reconsider.

### Pros of NEP 29

The obvious pro of NEP 29 is the ability to adopt language features added in newer versions of Python. Of particular interest are much simpler type hints added in Python 3.9. I don't see much of interest in 3.8 but I've been counting down the days until we can switch to 3.9+.

Another advantage is that our CI would be significantly simpler. We could remove all the `requirements/*.old` files that were added to avoid dependabot headaches. It would also mean 3 fewer CI runs (better for the environment), although I've honestly been thinking about only testing the newest and oldest versions of Python (currently 3.7 and 3.10), I don't see why testing 3.8 and 3.9 would be needed.

### Cons of NEP 29

Google Colab still uses Python 3.7. This is a big one for us. For libraries like numpy or PyTorch, it isn't hard to tell users to use an older release. But for TorchGeo, we're still very much in beta with new features and bug fixes being added every week. I'm not sure if we can afford to say "use an old release" to all Colab users.

Related to #413, @weiji14 may be interested in this discussion.

**Comment by adamjstewart** [comment]
@calebrob6 what version of Python does the Planetary Computer use?

**Reply by calebrob6** [reply]
3.8

**Comment by calebrob6** [comment]
I just looked into Colab a bit and it seems that you _can_ update the Python version in your runtime via `apt-get` if you really want to.

Generally, I'm for whatever makes our lives simpler. You are already spending a ton of your part-time effort on fixing broken CI and reducing that burden such that we have time to focus on useful features, benchmarking, etc. seems like a good move.

**Comment by adamjstewart** [comment]
Also relevant: https://github.com/googlecolab/colabtools/issues/2165

**Reply by adamjstewart** [reply]
They finally upgraded to 3.8! https://github.com/googlecolab/colabtools/issues/3246

**Comment by adamjstewart** [comment]
PyTorch just dropped Python 3.7 support: https://github.com/pytorch/pytorch/pull/93155

**Comment by adamjstewart** [comment]
Old discussion, but note that [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html) (3.5 year support schedule) has been superseded by [SPEC 0](https://scientific-python.org/specs/spec-0000/) (3 year support schedule). It sometimes takes PyTorch a few months to make a release with support for the latest Python version, but TorchGeo usually supports 3 versions of Python at a time.
